design/merge-refactor.md

```diff
@@ -0,0 +1,26 @@
+# Design: Refactor the Merge Operation
+
+## Problem Description
```

Comment:
general note about formatting. can you keep a blank line after the header.
https://google.github.io/styleguide/docguide/style.html#add-spacing-to-headings